### PR TITLE
#495 Add functionality for extending foundation-mda

### LINKED ISSUE #495: Feature: Add functionality for extending foundation-mda so that I can modify existing metamodel schemas in a downstream project
## Description
[Fermenter Issue #78](https://github.com/TechnologyBrewery/fermenter/issues/78) fixed an issue with extending a projects existing metamodel repository. This new functionality needs to be incorporated into aiSSEMBLE, so that downstream projects can extend our metamodel repository.

Currently within aiSSEMBLE we get the metamodel repository with the following method call:
```
AIOpsModelInstanceRepostory metamodelRepository = ModelInstanceRepositoryManager
                .getMetamodelRepository(AIOpsModelInstanceRepostory.class);
```
This prevents downstream users from being able to extend our `AIOpsModelInstanceRepostory`, see [Fermenter Issue #78](https://github.com/TechnologyBrewery/fermenter/issues/78) for more details. With the new functionality added to Fermenter to resolve that issue, we need to replace our existing calls to get the metamodel repository with the following new method:
```
AIOpsModelInstanceRepostory metamodelRepository = generationContext.getModelInstanceRepository();
```

## DOD
- [x] Update all calls to get the metamodel repository within generator classes to use the new method
  - [x] Use the `GenerationContext` object which is available in all generators
- [x] Check for compilation warnings within build output of `foundation-mda` for issues that might cause bugs when extending classes
- [x] Update `mda-maven-plugin` to handle an extended metamodel repository
- [x] Rename and add to Release Notes:
  - [x] `AIOpsModelInstanceRepostory` :arrow_right: `AissembleModelInstanceRepository`
  - [x] `AiopsMdaJsonUtils` :arrow_right: `AissembleMdaJsonUtils`
- [x] Increase testing in `foundation-mda`
  - [x] Create new unit test for `extending foundation-mda` 
- [x] Create way in Fermenter to get metamodel repository without direct access to the generation context (i.e. within a `MetamodelElement`)
  - [x] Release next minor version of Fermenter

## Test Strategy/Script
- **OTS Only:**
  - Run the following on the [78-enable-metamodel-extension](https://github.com/TechnologyBrewery/fermenter/tree/78-enable-metamodel-extension) branch in Fermenter: `mvn clean install`
- Run the following within aiSSEMBLE:
```
mvn clean install -pl :foundation-mda,:mda-maven-plugin -am -Dmaven.build.cache.skipCache
```

## References/Additional Context
[Fermenter Issue #78](https://github.com/TechnologyBrewery/fermenter/issues/78)

### PR DESCRIPTION
**Pull request by carter-cundiff**
All the changes generator classes are an update to use the new Fermenter functionality for getting the metadata repository from the generation context. This is to avoid harcdoding our Metamodel Repository class name in, which was preventing downstream projects from extending `foundation-mda`.
Old way:
```
AIOpsModelInstanceRepostory metamodelRepository = ModelInstanceRepositoryManager
                .getMetamodelRepository(AIOpsModelInstanceRepostory.class);
```
New way:
```
AissembleModelInstanceRepository metamodelRepository = (AissembleModelInstanceRepository) 
               context.getModelInstanceRepository();
```

Note in some metamodel classes (ex: [RecordFieldTypeElement.java](https://github.com/boozallen/aissemble/blob/dev/foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/RecordFieldTypeElement.java)) we don't have access to the generation context as it hasn't been created yet, so we instead have to use the Fermenter `ModelInstanceRepositoryManager` with new functionality for handling class extensions: [Fermenter PR](https://github.com/TechnologyBrewery/fermenter/pull/80)

Unit test files of interest:
- `foundation-mda`:
  - [Feature file](https://github.com/boozallen/aissemble/pull/505/files#diff-195a590121a10c73ffff9a5dfe54a11827827ba78f167c5504000f1a1c6776bf)
  - [Implementation Steps](https://github.com/boozallen/aissemble/pull/505/files#diff-2842f0014dc460048bfc9b321f8a1604d2b53bb20913b1b5b7c5a2406da93e06)
- `mda-maven-plugin`
  - [Feature file](https://github.com/boozallen/aissemble/pull/505/files#diff-f416da5aef6abc9e186c2ba3bf1f8e3c12f206021b64621fc9f953e99e5eb39f)
  - [Implementation steps](https://github.com/boozallen/aissemble/pull/505/files#diff-cfb1b8bea65f5119fbee1db9c4a57072a1724a00c5734224e43580919ee97707)

### CODE REVIEW

File: build-parent/pom.xml [hunk 1]
```diff
@@ -32,7 +32,7 @@
         <version.buildhelper.plugin>3.6.0</version.buildhelper.plugin>
         <version.buildnumber.plugin>3.2.1</version.buildnumber.plugin>
         <version.failsafe.plugin>${version.maven.surefire.plugin}</version.failsafe.plugin>
-        <version.fermenter>2.10.3</version.fermenter>
+        <version.fermenter>2.10.5</version.fermenter>
```
Comment (carter-cundiff) [inline build-parent/pom.xml:35]: I: This version of fermenter will be released prior to merging

File: foundation/foundation-maven-plugins/mda-maven-plugin/src/test/java/com/boozallen/mda/maven/MetamodelExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Given("a model instance repository extending foundation-mda")
+    public void a_model_instance_repository_extending_foundation_mda() {
+        this.modelRepositoryImpl = "com.boozallen.mda.maven.metadata.repository.ExtensionModelInstanceRepository";
+    }
+
+    @Given("a pipeline with metamodel extensions")
+    public void a_pipeline_with_metamodel_extensions() {
+        this.modelSource = new File("src/test/resources/models/model-extension.jar");
```
Comment (carter-cundiff) [inline foundation/foundation-maven-plugins/mda-maven-plugin/src/test/java/com/boozallen/mda/maven/MetamodelExtensionTest.java:60]: I: This jar file contains a pipeline with the new extension field.

**Review by d-ryan-ashcraft — APPROVED** [review]

File: foundation/foundation-mda/src/test/resources/specifications/extend-metamodel.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
+@extend-metamodel
+Feature: Extend foundation-mda so that you can modify existing metamodel schemas
+
+  Scenario Outline: Foundation-mda metamodel elements are extendable
+    Given a class extending the "<type>" metamodel with new simple and complex fields
+    When the "<profile>" profile is generated
```
Comment (epwilkins) [inline foundation/foundation-mda/src/test/resources/specifications/extend-metamodel.feature:6]: A: This feels like it has the potential to clash and the profiles we've selected feel a little arbitrary.  It's extra coverage so I'm OK leaving it, but I'd also be OK with just making this _"When the extended models are read in"_ or something to that effect.

**Review by epwilkins — APPROVED** [review]

File: foundation/foundation-mda/src/test/resources/specifications/extend-metamodel.feature (hunk 3 above)
Comment (carter-cundiff) [reply foundation/foundation-mda/src/test/resources/specifications/extend-metamodel.feature:6]: The reason I added it was cause the error itself occurs when the generators are run, so felt like the tests we insufficient if we didn't include them.

File: foundation/foundation-maven-plugins/mda-maven-plugin/src/test/java/com/boozallen/mda/maven/MetamodelExtensionTest.java (hunk 2 above)
Comment (aaron-gary) [reply foundation/foundation-maven-plugins/mda-maven-plugin/src/test/java/com/boozallen/mda/maven/MetamodelExtensionTest.java:60]: Q: Is this jar file typically generated by Maven?

Comment (carter-cundiff) [reply foundation/foundation-maven-plugins/mda-maven-plugin/src/test/java/com/boozallen/mda/maven/MetamodelExtensionTest.java:60]: Correct. We're using it to mimic what maven would be providing the plugin in an actual execution

**Review by aaron-gary — APPROVED** [review]

File: build-parent/pom.xml (hunk 1 above)
Comment (carter-cundiff) [reply build-parent/pom.xml:35]: Release: https://github.com/TechnologyBrewery/fermenter/releases/tag/root-2.10.5

**Pull request merged**